File: Java/Key Concepts/Kadane'sAlgorithm.java
//Question : Print the Subarray With Maximum Sum

    //Kadane's Algorithm
    public static void maxSubarraySum3(int arr[]) {
        int cs = 0;
        int ms = 0;

        for(int i=0; i<arr.length; i++) {
            cs = cs + arr[i];
            if(cs < 0) {
                cs = 0;
            } 
            ms = Math.max(ms, cs);
        }

        System.out.println("max subarray sum is : " + ms);
    }
    public static void main(String args[]) {
        int arr[] = {1, -2, 6, -1, 3};
        maxSubarraySum1(arr); 
        maxSubarraySum2(arr);
        maxSubarraySum3(arr);
    }
}

File: 02-C and C++/C++/Data Structures/Arrays/1D Arrays/Kadane's_Algorithm.cpp
#include <iostream>
#include <algorithm> // For std::max
using namespace std;

// Kadane's Algorithm to find the maximum subarray sum
void maxSubarraySum3(int arr[], int size) {
    int currentSum = 0;
    int maxSum = 0;

    for(int i = 0; i < size; i++) {
        currentSum += arr[i];
        if (currentSum < 0) {
            currentSum = 0;
        }
        maxSum = max(maxSum, currentSum);
    }

    cout << "Max subarray sum is: " << maxSum << endl;
}

int main() {
    int arr[] = {1, -2, 6, -1, 3};
    int size = sizeof(arr)/sizeof(arr[0]);

    maxSubarraySum3(arr, size); 

    return 0;
}
